doc/adrs/157_improve_react_component_representations_rendering.adoc

```diff
@@ -0,0 +1,175 @@
+
+
+We don't have to use as many types of changes as the one used in ReactFlow of course and could just send only the list of nodes concerned by a change, I'm using the ones from ReactFlow to highlight the fact that we are already manipulating only the concerned nodes in some cases.
+
+
+===== Live incremental changes
+
+
+When two users are working on a diagram, if a user is moving a node the others are not seeing the node moving.
+
+
+By sending events given by ReactFlow (like ```NodePositionChange```) when dragging a node to the back-end, these changes could then be propagated to the other users subscribed to the diagram.
+
+
+This would allow more feedback when working with several users on the same diagram.
+
+
+But these changes might be invalidated if the user is dropping the node into a wrong position, in this case it might be harder to keep the diagrams in sync.
+
+
+=== Layout
+
+
+One thing to also note is that we have an ```initial layouting phase``` that does a lot of things that we already do in ```handleNodesChange``` like setting the side of the handles.
+
+
+== Explorer representation
+
+
+The only representation that I think would also benefit from a similar approach is the explorer.
+
+
+Since the explorer can have thousands of elements, changing the name of a ```TreeItem``` should not result in a rerender of all ```TreeItem```.
+
+
+For that, 2 approaches can be used
+
+
+=== Store TreeItems data in a store.
+
+
+With a ```store``` each component TreeItem would ```subscribe``` with a ```selector``` to the corresponding data object, if the ```reference``` to the object changed in the store then the treeItem would rerender.
+
+
+We could send an event like ```TreeItemReset``` that would change only a specific treeItem in the store and thus re-render only this treeItem (since it would be the only one subscribed to this change with the selector).
+
+
+In the same idea, ```TreeItemExpended``` event would only send the child of the expended TreeItem.
+
+
+Adding a store would be a big change, but we already use the ```internal store of ReactFlow``` and it has proven very useful to re-render specific components.
+
+
+=== Store TreeItems in a state (like we already do).
+
+
+We could do a similar approach without a store by keeping the Tree in a ```state``` and only modify the concerned TreeItem, this would reduce re-render a lot for big trees but all the parent elements of a TreeItem concerned by a change would have to be rerendered.
+
+
+We should start without a store and see if also re rendering parents impacts performance a lot or not.
+
+
+== Decision
+
+
+We should start by comparing the payloads when the client receives and update only the revelant React Component.
+
+Then measure the performance improvements before trying a more complex approach.
```

Comment:
I can understand in the grand scheme of things how we could use a solution like that but I don't see what would motive it right now.

> The changes sended by the back-end could look like the changes used by the ```ReactFlow renderer``` meaning for the nodes

That would be a massive change in itself. While it would certainly help with minor operations like moving a node, most of what is possible in a diagram has a sizable impact, potentially changing dozens of elements in a diagram. I'm not even sure that "computing the incremental change" would be faster than just sending the new diagram. I'm not sure which performance problem identified RIGHT NOW you would solve with this.

Our approach to performance improvements will be guided by concrete performance problems identified right now. This ADR involves multiples massive architectural changes on both the backend and the frontend of multiple parts of the code. It may make some good points but it's way too vague and generic.

If you want to move forward with such proposal, when you will have time affected to performance improvement or during a cool down:

- Find and measure a specific performance problem that affects our end users
- Come up with an ADR to solve that specific problem

We could move forward one day with incremental changes sent over the subscription for representations, we could move forward with other kind of changes for tree based representations too but in this state, I don't understand what clear issue this ADR is trying to solve and how it would solve it. The constraints we have one tree based representations and diagram or form based representations are very different and the solution would have to be tailor made for each use case. I feel that this document should probably be 2 or 3 different ADRs but I don't clearly see the performance issue identified for each use case.